### DISCUSSION TITLE
llama-server cache_prompt

### DISCUSSION THREAD

**Discussion by drunnells** [open-webui/open-webui#7137] [Ideas]
I have been using open-webui to connect to my llama-server instance. It's working great, but with long contexts it becomes quite slow to continue a chat conversation. Long chats with this setup can be significantly improved with the "cache_prompt" parameter passed in with the completion payload. See:

https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md

`cache_prompt`: Re-use KV cache from a previous request if possible. This way the common prefix does not have to be re-processed, only the suffix that differs between the requests. Because (depending on the backend) the logits are **not** guaranteed to be bit-for-bit identical for different batch sizes (prompt processing vs. token generation) enabling this option can cause nondeterministic results. Default: `false`

I've made this is a minor change to backend/open_webui/utils/payload.py to include it with the apply_model_params_to_body_openai mappings and AdvancedParams.svelte to have a configuration toggle for the model:

![Screenshot 2024-11-20 at 9 44 39 PM](https://github.com/user-attachments/assets/e6b90ed2-df54-4c56-ba46-aad9c7b6132e)

I'd love to fork and make a PR for this as the change seems to be working as expected for me. Just last week there was some discussion in llama.cpp's github that specifically called out that open-webui doesn't support this: https://github.com/ggerganov/llama.cpp/discussions/10311

**Comment by drunnells** [comment]
I've created the PR fr this change: https://github.com/open-webui/open-webui/pull/7237
